test/integration/ddl_sharding_key_test.lua

```diff
@@ -0,0 +1,181 @@
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+end
+
+-- Right now CRUD's plan for select doesn't support sharding key and it leads
+-- to map reduce (select on all replicasets). To avoid map-reduce one need to
+-- add a separate index by field name, used in select's condition. We plan to
+-- fix this in scope of https://github.com/tarantool/crud/issues/213
+pgroup.test_select_wont_lead_map_reduce = function(g)
+    local space_name = 'customers_name_key_uniq_index'
+    local customers = helpers.insert_objects(g, space_name, {
+        {id = 1, name = 'Viktor Pelevin', age = 58},
+        {id = 2, name = 'Isaac Asimov', age = 72},
+        {id = 3, name = 'Aleksandr Solzhenitsyn', age = 89},
+        {id = 4, name = 'James Joyce', age = 59},
+        {id = 5, name = 'Oscar Wilde', age = 46},
+        {id = 6, name = 'Ivan Bunin', age = 83},
+        {id = 7, name = 'Ivan Turgenev', age = 64},
+        {id = 8, name = 'Alexander Ostrovsky', age = 63},
+        {id = 9, name = 'Anton Chekhov', age = 44},
+    })
+    t.assert_equals(#customers, 9)
+
+    -- Disable vshard's rebalancer and account current statistics of SELECT
+    -- calls on storages before calling CRUD select. Rebalancer may screw up
+    -- statistics of SELECT calls, so we will disable it.
+    local servers = g.cluster.servers
+    local select_total_counter_before = 0
+    for n, _ in ipairs(servers) do
+        local c = g.cluster.servers[n].net_box:eval([[
+            local vshard = require('vshard')
+            vshard.storage.rebalancer_disable()
+            assert(vshard.storage.sync(2) == true)
+            assert(vshard.storage.rebalancing_is_in_progress() == false)
+
+            return box.stat().SELECT.total
+        ]])
+        select_total_counter_before = select_total_counter_before + c
+    end
+
+    -- Make a CRUD's SELECT.
+    local result, err = g.cluster.main_server.net_box:call('crud.select', {
+        space_name, {{'==', 'name', 'Anton Chekhov'}}
+    })
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+
+    -- Enable vshard's rebalancer and account current statistics of SELECT
+    -- calls on storages after calling CRUD select.
+    local select_total_counter_after = 0
+    for n, _ in ipairs(servers) do
+        local c = g.cluster.servers[n].net_box:eval([[
+            local vshard = require('vshard')
+            local stat = box.stat().SELECT.total
+            vshard.storage.rebalancer_enable()
+
+            return stat
+        ]])
+        select_total_counter_after = select_total_counter_after + c
+    end
+
+    -- Compare total counters of SELECT calls on cluster's storages before and
+    -- after calling SELECT on router. Make sure no more than 1 storage changed
+    -- SELECT counter. Otherwise we lead map reduce.
+    local diff = select_total_counter_after - select_total_counter_before
+    t.assert_le(diff, 4)
+    t.assert_ge(diff, 2)
```

Comment:
The diff is checked against the `2 <= diff <= 4` condition. But the comment says:

> Make sure no more than 1 storage changed SELECT counter.

I don't understand it. Whether you want to fix 'bad' behaviour here and fix it in #213? If so, leave a comment.

Why the value should be between 2 and 4? AFAIR, we have two replicasets here. If the check is correct, the value should be 2 for map-reduce.

Reply:
Test has been updated. See patch above and in a branch.
Short answer: `box.stat().SELECT.total` is not reliable as a source of storage statistics.